internal/utils.go

```diff
@@ -77,25 +77,46 @@ func (l *LockMap) Get(k string) string {
 // Implementation and related tests were taken/inspired by felixge/countermap
 // https://github.com/felixge/countermap/pull/2
 type XSyncMapCounterMap struct {
-	counts *xsync.MapOf[string, *xsync.Counter]
+	counts  [2]*xsync.MapOf[string, *xsync.Counter]
+	mu, muD sync.Mutex
+	index   int
 }
 
 func NewXSyncMapCounterMap() *XSyncMapCounterMap {
-	return &XSyncMapCounterMap{counts: xsync.NewMapOf[string, *xsync.Counter]()}
+	return &XSyncMapCounterMap{
+		counts: [2]*xsync.MapOf[string, *xsync.Counter]{
+			xsync.NewMapOf[string, *xsync.Counter](),
+			xsync.NewMapOf[string, *xsync.Counter](),
+		},
+	}
 }
 
 func (cm *XSyncMapCounterMap) Inc(key string) {
-	val, ok := cm.counts.Load(key)
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	i := cm.index
+	val, ok := cm.counts[i].Load(key)
 	if !ok {
-		val, _ = cm.counts.LoadOrStore(key, xsync.NewCounter())
+		val, _ = cm.counts[i].LoadAndStore(key, xsync.NewCounter())
 	}
 	val.Inc()
 }
 
+func (cm *XSyncMapCounterMap) swap() int {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	old := cm.index
+	cm.index = cm.index ^ 1
+	return old
+}
+
 func (cm *XSyncMapCounterMap) GetAndReset() map[string]int64 {
+	cm.muD.Lock()
+	defer cm.muD.Unlock()
 	ret := map[string]int64{}
-	cm.counts.Range(func(key string, _ *xsync.Counter) bool {
-		v, ok := cm.counts.LoadAndDelete(key)
+	i := cm.swap()
+	cm.counts[i].Range(func(key string, _ *xsync.Counter) bool {
```

Comment:
If we have the singleflight and swap, we don't need to `LoadAndDelete`, we can just read everything and delete it at one go https://github.com/puzpuzpuz/xsync/blob/800e3a0ceeab7d9a5c17df16241c1a4cca0da524/map.go#L729-L732

Also checkout https://github.com/puzpuzpuz/xsync/blob/800e3a0ceeab7d9a5c17df16241c1a4cca0da524/map.go#L207 and https://github.com/puzpuzpuz/xsync/blob/800e3a0ceeab7d9a5c17df16241c1a4cca0da524/mapof.go#L156